plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Globalify.hs

```diff
@@ -0,0 +1,75 @@
+{-# LANGUAGE LambdaCase   #-}
+{-# LANGUAGE TypeFamilies #-}
+module UntypedPlutusCore.Transform.Globalify (globalifyProgram, globalifyTerm, maxGlobal) where
+
+import           Control.Lens.Fold
+import           Control.Lens.Plated
+import           Control.Monad.Reader
+import           Data.Semigroup
+import           Data.Word
+import qualified PlutusCore.Name               as TPLC
+import           PlutusCore.Quote
+import           UntypedPlutusCore.Core.Plated
+import           UntypedPlutusCore.Core.Type
+import           UntypedPlutusCore.Rename
+import           UntypedPlutusCore.Subst
+
+{- Note [Globalifying]
+It's a bit awkward to deal with global variables. Initially, I tried to match on things of the shape
+
+    [ (lam n body) rhs ]
+
+i.e. things that look like a compiled let-binding. But this is not sufficient, because we generate code like
+this from datatype bindings:
+
+    [ [ [ (lam n1 (lam n2 (lam n3 body) ) ) rhs3 ] rhs2 ] rhs1]
+
+Furthermore, scripts as we see them before evaluation will be applied to some number of arguments!
+
+I spent a while trying to figure out how to identify the stuff that we wanted as globals, and then I thought
+of a different way of thinking of it.
+
+> Look for lambdas which are only ever called once
```

Comment:
In that case, wouldn't you want to inline them?  Maybe we can't do that because we won't know some things until the script is applied to its arguments on the chain?

Reply:
There's a slight difference:

```
[ (lam g ( ... g .... g .. g)) rhs ]
```
vs
```
[ (lam w ... ) [ (lam single (... single ...)) ...] ]
```

In the first one, the lambda only gets *applied* once, so we can use a global variable for it. But the variable might get used several times, so we don't necessarily want to do inlining.

In the second one, the lambda may get applied lots of times (since it's on the RHS of an application), but the variable it binds is used only once, so if we *can* inline it, it's good.

It's the difference between whether the lambda is applied once versus whether the bound variable is used once.